### DISCUSSION TITLE
Intercept requests to add data when using as a framework with JS

### DISCUSSION THREAD

**Discussion by PaulDotterer** [pocketbase/pocketbase#3142] [Q&A] [answered]
Hello all,

i have been experimenting with using pocketbase as a framework since the new addition of JS in v0.17.0.
I cant figure out how to append certain data to the request record object in the on `onRecordBeforeUpdateRequest` or `onRecordBeforeCreateRequest` hooks.
As an example:

i have a collection with this fields:

"blog_entry"
- subject | text
- content | text
- createdBy | relation on users collection
- updatedBy | relation on users collection

i would like to automatically set createdBy to the user record before the resource is persisted and the updatedBy field every time a change to the record occurs.

Updating after the request is not ideal, because this triggers an unnecessary second SSE event.

Currently im populating these field with serverside logic before the request gets send, but i would love to implement it directly with pocketbase to have a uniform solution.

I have tried modifying the request object but i always get an error, if there is a way to do this and im missing something, please let me know!

Function i tried:

```
onRecordBeforeUpdateRequest(
	(e) => {
		const info = $apis.requestInfo(e.httpContext);
                e.record.updatedBy = info.authRecord.id
                return e
	},
	"blog_entry"
);
```

**Comment by ganigeorgiev** [comment]
The above example seems correct but with 2 minor details:
- `info.authRecord` could be `null` if the request was sent by admin or non-authenticated clients, so you can try changing it to `info.authRecord?.id`
- the hook return result could be either `null` or a Go error value, aka. remove `return e`

**Reply by PaulDotterer** [reply]
Thanks for the fast reply!
I tried your suggestion but get the following error:

Cannot assign to property updatedBy of a host object at <eval>:3:11(15)

The function now looks like this:

```
onRecordBeforeCreateRequest((e) => {
	const info = $apis.requestInfo(e.httpContext);
	e.record.updatedBy = info.authRecord?.id;
}, "blog_entry");

```

**Reply by ganigeorgiev** [reply, marked as answer]
Oh, sorry. The custom record schema fields must be set using the `set()` method, aka. `e.record.set("updatedBy", info.authRecord?.id)`.

**Reply by PaulDotterer** [reply]
Thank you so much!
Worked like a charm :)
Did i miss this in the documentation?

**Reply by ganigeorgiev** [reply]
You can find list with the available record helper methods in https://pocketbase.io/docs/js-records/#getset-record-fields.

The setter is needed because `e.record` is the underlying wrapped Go struct and we can't bind dynamic props to it.

_Eventually in the future this may not be necessary if we try to overwrite the goja js prototype object and replace it with a `Proxy` that will take care for the getters and setters but it could have some unintended side-effects as all wrapped objects share the same prototype and I've left it aside for now._

**Reply by PaulDotterer** [reply]
Thank you very much!
